pthread/pthread.c

```diff
@@ -51,6 +55,25 @@ static const pthread_attr_t pthread_attr_default = {
 };
 
 
+static void pthread_ctx_get(pthread_ctx *ctx)
+{
+	mutexLock(pthread_common.pthread_list_lock);
+	ctx->refcount++;
+	mutexUnlock(pthread_common.pthread_list_lock);
+}
+
+
+static void pthread_ctx_put(pthread_ctx *ctx)
+{
+	mutexLock(pthread_common.pthread_list_lock);
+	ctx->refcount--;
+	if (ctx->refcount == 0) {
+		free(ctx);
+	}
+	mutexUnlock(pthread_common.pthread_list_lock);
```

Comment:
So we don't hold mutex during free - if refcount <= 0 only we hold this pointer and it has been removed from the list, so it's safe to do so.
```suggestion
	int refcnt;
	
	mutexLock(pthread_common.pthread_list_lock);
	refcnt = --ctx->refcount;
	mutexUnlock(pthread_common.pthread_list_lock);
	
	if (refcnt <= 0) {
		free(ctx);
	}
```